support/testsupport/src/main/java/com/hotels/styx/support/LatchesAndBarriers.java

```diff
@@ -0,0 +1,57 @@
+/*
+  Copyright (C) 2013-2018 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.support;
+
+import java.util.concurrent.BrokenBarrierException;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static java.lang.String.format;
+import static java.lang.Thread.currentThread;
+
+/**
+ * Makes `CountDownLatch` more convenient for unit tests.
+ */
+public final class LatchesAndBarriers {
+    private LatchesAndBarriers() {
+    }
+
+    public static void await(CountDownLatch latch, long timeout, TimeUnit unit) throws IllegalStateException {
+        try {
+            boolean completed = latch.await(timeout, unit);
+
+            if (!completed) {
+                throw new IllegalStateException(format("Latch timed out: max wait was %s %s", timeout, unit));
+            }
+        } catch (InterruptedException e) {
+            currentThread().interrupt();
```

Comment:
Surely we need to re-throw  something from the InterruptedException case? Otherwise `await` just returns and the calling code carries on as if nothing happened.

Reply:
Sure I failed to notice that we had stopped throwing an Exception here. My advice is that we just let InterruptedException propagate...

Reply:
Unfortunately, `Observable.subscribe` takes an `Action1` as its argument, which will not allow `InterruptedException` to be thrown.